File: java_src/foam/dao/EnabledAwareDAO.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.dao;

import foam.mlang.predicate.Predicate;
import foam.mlang.order.Comparator;
import foam.dao.Sink;

public class EnabledAwareDAO
  extends ProxyDAO
{
  public static final Predicate PREDICATE = foam.mlang.MLang.EQ(EnabledAware.ENABLED, true);

  public Sink select_(Sink s, Integer skip, Integer limit, Comparator order, Predicate predicate) {
    return super.select_(s, skip, limit, order, PREDICATE);
  }

  public void removeAll(Integer skip, Integer limit, Comparator order, Predicate predicate) {
    super.removeAll(skip, limit, order, PREDICATE);
  }
}

File: test/src/dao/EnabledAwareDAO.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

describe('EnabledAwareDAO', function() {
  var Entity;
  var dao_;

  beforeEach(function() {
    foam.CLASS({
      package: 'foam.dao.test',
      name: 'Entity',

      properties: [
        'id',
        'enabled'
      ]
    });
    Entity = foam.dao.test.Entity;
    dao_ = foam.dao.EnabledAwareDAO.create({
      of: Entity,
      delegate: foam.dao.ArrayDAO.create({ of: Entity })
    });
  });

  it('should not find disabled entity', function() {
    dao_.put(Entity.create({ id: 1, enabled: false }));
    dao_.find(1).then(function(o) {
      expect(o).toBeNull();
    });
  });

  it('should not include disabled entity on select', function() {
    dao_.put(Entity.create({ id: 1, enabled: false }));
    dao_.put(Entity.create({ id: 2, enabled: true }));
    dao_.select().then(function(db) {
      expect(db.array.length).toEqual(1);
      expect(db.array[0].id).toEqual(2);
    });
  });
});
